WalletWasabi/Tor/Socks5/TorSocks5ClientPool.cs

```diff
@@ -0,0 +1,303 @@
+			{
+				return await ClearnetHttpClient.SendAsync(request, cancellationToken).ConfigureAwait(false);
+			}
+
+			int i = 0;
+			int attemptsNo = 3;
+
+			try
+			{
+				do
+				{
+					i++;
+					IPoolItem poolItem = await ObtainPoolItemAsync(request, isolateStream, cancellationToken).ConfigureAwait(false);
+					IPoolItem? itemToDispose = poolItem;
+
+					try
+					{
+						Logger.LogTrace($"['{poolItem}'] About to send request.");
+						HttpResponseMessage response = await SendCoreAsync(poolItem.GetTransportStream(), request, cancellationToken).ConfigureAwait(false);
+
+						// Client works OK, no need to dispose.
+						itemToDispose = null;
+
+						// Let others use the client.
+						var state = poolItem.Unreserve();
+						Logger.LogTrace($"['{poolItem}'] Unreserve. State is: '{state}'.");
+
+						TorDoesntWorkSince = null;
+						LatestTorException = null;
+
+						return response;
+					}
+					catch (TorConnectCommandException e) when (e.RepField == RepField.TtlExpired)
+					{
+						// If we get TTL Expired error then wait and retry again linux often does this.
+						Logger.LogTrace(e);
+
+						await Task.Delay(1000, cancellationToken).ConfigureAwait(false);
+
+						if (i == attemptsNo)
+						{
+							Logger.LogDebug($"All {attemptsNo} attempts failed.");
+							throw;
+						}
+					}
+					catch (SocketException e) when (e.ErrorCode == (int)SocketError.ConnectionRefused)
+					{
+						Logger.LogTrace(e);
+						throw new TorConnectionException("Connection was refused.", e);
+					}
+					catch (Exception e)
+					{
+						Logger.LogTrace(e);
+						throw;
+					}
+					finally
+					{
+						(itemToDispose as IDisposable)?.Dispose();
+					}
+				} while (i < attemptsNo);
+			}
+			catch (Exception ex)
+			{
+				OnTorRequestFailed(ex);
+				throw;
+			}
+
+			throw new NotImplementedException("This should never happen.");
+		}
+
+		private async Task<IPoolItem> ObtainPoolItemAsync(HttpRequestMessage request, bool isolateStream, CancellationToken token)
+		{
+			Logger.LogTrace($"> request='{request.RequestUri}', isolateStream={isolateStream}");
+
+			string host = GetRequestHost(request);
+
+			do
+			{
+				using (await ObtainPoolItemAsyncLock.LockAsync(token).ConfigureAwait(false))
+				{
+					(bool canBeAdded, IPoolItem? poolItem) = PoolItemManager.GetPoolItem(host, isolateStream);
+
+					if (poolItem is { })
+					{
+						Logger.LogTrace($"[OLD {poolItem}]['{request.RequestUri}'] Re-use existing Tor SOCKS5 connection.");
+						return poolItem;
+					}
+
+					if (canBeAdded)
+					{
+						poolItem = await CreateNewPoolItemNoLockAsync(request, isolateStream, token).ConfigureAwait(false);
+
+						if (poolItem is { })
+						{
+							Logger.LogTrace($"[NEW {poolItem}]['{request.RequestUri}'] Using new Tor SOCKS5 connection.");
+							return poolItem;
+						}
+					}
+				}
+
+				Logger.LogTrace("Wait 1s for a free pool item.");
+				await Task.Delay(1000, token).ConfigureAwait(false);
+			} while (true);
```

Comment:
! token.IsCancelle

For now, it does not make sense but cleaner for the future.

Reply:
I would need to throw an exception then.